src/category_theory/sites/sieves.lean

```diff
@@ -0,0 +1,250 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, E. W. Ayers
+-/
+
+import category_theory.over
+import category_theory.limits.shapes.finite_limits
+import category_theory.yoneda
+import order.complete_lattice
+import data.set.lattice
+
+/-!
+# Theory of sieves
+
+- For an object `X` of a category `C`, a `sieve X` is a set of morphisms to `X`
+  which is closed under left-composition.
+- The complete lattice structure on sieves is given, as well as the Galois insertion
+  given by downward-closing.
+- A `sieve X` (functorially) induces a presheaf on `C` together with a monomorphism to
+  the yoneda embedding of `X`.
+
+## Tags
+
+sieve, pullback
+-/
+
+universes v u
+namespace category_theory
+
+/--
+For an object `X` of a category `C`, a `sieve X` is a set of morphisms to `X` which is closed under
+left-composition.
+-/
+structure sieve {C : Type u} [category.{v} C] (X : C) :=
+(arrows : Π {Y}, set (Y ⟶ X))
+(downward_closed : ∀ {Y Z f} (hf : arrows f) (g : Z ⟶ Y), arrows (g ≫ f))
+attribute [simp, priority 100] sieve.downward_closed
+
+namespace sieve
+
+variables {C : Type u} [category.{v} C]
+
+variables {X Y Z : C} {S R : sieve X}
+
+/-- A sieve gives a subset of the over category of `X`. -/
+def get_arrow_set (S : sieve X) : set (over X) := λ f, S.arrows f.hom
+
+lemma arrows_ext : Π {R S : sieve X}, R.arrows = S.arrows → R = S
+| ⟨Ra, _⟩ ⟨Sa, _⟩ rfl := rfl
+
+protected lemma ext {R S : sieve X}
+  (h : ∀ ⦃Y⦄ (f : Y ⟶ X), R.arrows f ↔ S.arrows f) :
+  R = S :=
+arrows_ext $ funext $ λ x, funext $ λ f, propext $ h f
+
+protected lemma ext_iff {R S : sieve X} :
+  R = S ↔ (∀ ⦃Y⦄ (f : Y ⟶ X), R.arrows f ↔ S.arrows f) :=
+⟨λ h Y f, h ▸ iff.rfl, sieve.ext⟩
+
+open lattice
+
+/-- The supremum of a collection of sieves: just the union of them all. -/
+protected def Sup (𝒮 : set (sieve X)) : (sieve X) :=
+{ arrows := λ Y, {f | ∃ S ∈ 𝒮, sieve.arrows S f},
+  downward_closed := λ Y Z f, by { rintro ⟨S, hS, hf⟩ g, exact ⟨S, hS, S.downward_closed hf _⟩ } }
+
+/-- The infimum of a collection of sieves: the intersection of them all. -/
+protected def Inf (𝒮 : set (sieve X)) : (sieve X) :=
+{ arrows := λ Y, {f | ∀ S ∈ 𝒮, sieve.arrows S f},
+  downward_closed := λ Y Z f hf g S H, S.downward_closed (hf S H) g }
+
+/-- The union of two sieves is a sieve. -/
+protected def union (S R : sieve X) : sieve X :=
+{ arrows := λ Y f, S.arrows f ∨ R.arrows f,
+  downward_closed := by { rintros Y Z f (h | h) g; simp [h] } }
+
+/-- The intersection of two sieves is a sieve. -/
+protected def inter (S R : sieve X) : sieve X :=
+{ arrows := λ Y f, S.arrows f ∧ R.arrows f,
+  downward_closed := by { rintros Y Z f ⟨h₁, h₂⟩ g, simp [h₁, h₂] } }
+
+/--
+Sieves on an object `X` form a complete lattice.
+We generate this directly rather than using the galois insertion for nicer definitional
+properties.
+-/
+instance : complete_lattice (sieve X) :=
+{ le           := λ S R, ∀ Y (f : Y ⟶ X), S.arrows f → R.arrows f,
```

Comment:
One trick to have nice defeqs and use the gi, is to use `complete_lattice.copy`. So you first get a lattice structure via the gi, and then copy all the proofs, but replace the data to provably equal things with nice defeq properties.
I don't say that you should use that trick here (but I also don't say you shouldn't :grinning:).

Reply:
I think I'll leave it as is unless there's a compelling reason to switch?